app/src/test/js/keysCtrlTest.js

```diff
@@ -0,0 +1,158 @@
+                expect(scope.crumbs[1]).toEqual(response.tree.children[0]);
+                expect(scope.parent).toEqual(response.tree.children[0]);
+
+                scope.manageChildren(scope.parent.children[0]);
+                expect(scope.crumbs.length).toEqual(3);
+                expect(scope.crumbs[0]).toEqual(response.tree);
+                expect(scope.crumbs[1]).toEqual(response.tree.children[0]);
+                expect(scope.crumbs[2]).toEqual(
+                        response.tree.children[0].children[0]);
+                expect(scope.parent).toEqual(
+                        response.tree.children[0].children[0]);
+
+                scope.gotoCrumb(scope.crumbs[0]);
+                expect(scope.crumbs.length).toEqual(1);
+                expect(scope.crumbs[0]).toEqual(response.tree);
+                expect(scope.parent).toEqual(response.tree);
+            }
+
+            it('crumb and manage chidren tests', function() {
+                crumbTest(syncKeysScope, syncKeysSuccessResponse);
+                crumbTest(searchKeysScope, searchKeysSuccessResponse);
+            });
+
+            function addTest(scope, response, name, type, value) {
+                expect(scope.addMode).toEqual(false);
+                expect(scope.crumbs.length).toEqual(1);
+                scope.addChild(scope.parent);
+                expect(scope.addMode).toEqual(true);
+                expect(scope.crumbs.length).toEqual(2);
+                scope.addFormName = name;
+                scope.addFormValue = value;
+                scope.requestAdd();
+                http.expectPOST(
+                        "api/key/addChild/" + scope.parent.id + "?name=" + name
+                                + "&type=" + type + "&value=" + value).respond(
+                        response);
+                http.flush();
+                expect(scope.parent).toEqual(response.result);
+                expect(scope.entities || scope.searchKeys).toEqual(
+                        response.tree);
+            }
+
+            it('add tests', function() {
+                addTest(syncKeysScope, syncKeysSuccessResponse, "testName",
+                        "Entity", "testValue");
+                addTest(searchKeysScope, searchKeysSuccessResponse,
+                        "SearchKeys", "SearchKey", "testValue");
+            });
+
+            function updateTests(scope, response, name, type, value) {
+                expect(scope.updateMode).toEqual(false);
+                expect(scope.crumbs.length).toEqual(1);
+                scope.editChild(scope.parent.children[0]);
+                expect(scope.updateMode).toEqual(true);
+                expect(scope.crumbs.length).toEqual(2);
+                expect(scope.parent).toEqual(response.tree.children[0]);
+                scope.parent.name = name;
+                scope.parent.value = value;
+                scope.requestUpdate();
+                http.expectPUT(
```

Comment:
Another consideration is that you are also testing the service layer here (`EsbMessageService`). You could mock out this service and just test the controllers are using it correctly. Then, ideally you would have separate tests for `EsbMessageService` that mock out `$http` as you are doing here. This way you get the advantage of the service layer in your tests ... should the backend change, you just update your service tests, because only your service has changed, as opposed to updating controller tests where only a service has changed. In other words, you'd be taking advantage of dependency injection.

Admittedly, it doesn't look it is totally straightforward to stub an asynchronous service. You can inject $provide, and then you can do something like `$provide.('EsbMessageService', mockEsbMessageService);` where `mockEsbMessageService` is (spied on)[http://jasmine.github.io/2.0/introduction.html#section-Spies]. The trouble is setting up what the callbacks should be called with. You could define it in your "stub" service implementation, but then it is far away from the tests. You'd also have to add a `flush` method on the service. [Async programming is kind of a pain in the butt.](https://www.usenix.org/legacy/events/hotos03/tech/full_papers/vonbehren/vonbehren_html/index.html)